spec/advance_enumerable_spec.rb

```diff
@@ -0,0 +1,163 @@
+      expect { 1.my_each }.to raise_error(NoMethodError)
+    end
+  end
+
+  describe '#my_each_with_index' do
+    it 'performs operation on each element of array if block is given' do
+      expect(arr.my_each_with_index { |x, n| puts x, n }).to eql(arr.each_with_index { |x, n| puts x, n })
+    end
+    it 'returns to_enum if no block is given' do
+      expect(arr.my_each_with_index).to be_a(Enumerator)
+    end
+    it 'returns an error if an argument is given' do
+      expect { arr.my_each_with_index(1) }.to raise_error(ArgumentError)
+    end
+    it 'returns an error if something else is given instead of an array' do
+      expect { 1.my_each_with_index }.to raise_error(NoMethodError)
+    end
+  end
+
+  describe '#my_select' do
+    it 'returns an array of items with satisfies condition' do
+      expect(arr.my_select { |x| x < 4 }).to eql(arr.select { |x| x < 4 })
+    end
+    it 'returns to_enum if no block is given' do
+      expect(arr.my_select).to be_a(Enumerator)
+    end
+    it 'returns an error if an argument is given' do
+      expect { arr.my_select(1) }.to raise_error(ArgumentError)
+    end
+    it 'returns an error if something else is given instead of an array' do
+      expect { 1.my_select }.to raise_error(NoMethodError)
+    end
+  end
+
+  describe '#my_all?' do
+    it 'returns true if all the items of array satisfies the condition' do
+      expect(arr.my_all? { |x| x < 1 }).to eql(arr.all? { |x| x < 1 })
+    end
+    it 'returns true if pattern is matched or array has all elements same as argument' do
+      expect(arr.my_all?(/d/)).to eql(arr.all?(/d/))
+    end
+    it 'returns true if all the elements are true if we dont pass block and argument' do
+      expect(arr.my_all?).to be true
+    end
+    it 'returns true if an empty array is given' do
+      expect([].my_all?).to be true
+    end
+    it 'returns an error if something else is given instead of an array' do
+      expect { 1.my_all? }.to raise_error(NoMethodError)
+    end
+  end
+
+  describe '#my_none?' do
+    it 'returns true if none of the items of array satisfies the condition' do
+      expect(arr.my_none? { |x| x < 1 }).to eql(arr.none? { |x| x < 1 })
+    end
+    it 'returns true if none of the items match the pattern passed as argument' do
+      expect(arr.my_none?(/s/)).to eql(arr.none?(/s/))
+    end
+    it 'returns true if none of the elements are true if we dont pass block and argument' do
+      expect(arr_nil.my_none?).to be true
+    end
+    it 'returns true if an empty array is given' do
+      expect([].my_none?).to be true
+    end
+    it 'returns an error if something else is given instead of an array' do
+      expect { 1.my_none? }.to raise_error(NoMethodError)
+    end
+  end
+
+  describe '#my_any?' do
+    it 'returns true if any of the items of array satisfies the condition' do
+      expect(arr.my_any? { |x| x < 1 }).to eql(arr.any? { |x| x < 1 })
```

Comment:
You should have tests to check if your method works on other types of enumerables besides arrays.